server/src/main/java/io/crate/execution/dml/DynamicIndexer.java

```diff
@@ -116,27 +114,84 @@ public void indexValue(Object value,
             onDynamicColumn.accept(newColumn);
         }
         value = type.sanitizeValue(value);
+        indexer.collectSchemaUpdates(
+            value,
+            onDynamicColumn,
+            synthetics
+        );
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void indexValue(Object value,
+                           XContentBuilder xcontentBuilder,
+                           Consumer<? super IndexableField> addField,
+                           Map<ColumnIdent, Synthetic> synthetics,
+                           Map<ColumnIdent, ColumnConstraint> toValidate) throws IOException {
+        if (type == null) {
+            // At the second phase of indexing type is not null in almost all cases
+            // except the case with array of nulls
+            type = guessType(value);
+        }
+        StorageSupport<?> storageSupport = type.storageSupport();
+        if (storageSupport == null) {
+            if (handleEmptyArray(type, value, xcontentBuilder)) {
+                type = null; // guess type again with next value
+                return;
+            }
+            throw new IllegalArgumentException(
+                "Cannot create columns of type " + type.getName() + " dynamically. " +
+                "Storage is not supported for this type");
+        }
+        boolean nullable = true;
+        Symbol defaultExpression = null;
+        Reference newColumn = new SimpleReference(
+            refIdent,
+            RowGranularity.DOC,
+            type,
+            ColumnPolicy.DYNAMIC,
+            IndexType.PLAIN,
+            nullable,
+            storageSupport.docValuesDefault(),
+            position,
+            COLUMN_OID_UNASSIGNED,
+            false,
+            defaultExpression
+        );
+        if (indexer == null) {
+            // Reuse indexer if phase 1 already created one.
+            // Phase 1 mutates indexer.innerTypes on new columns creation.
+            // Phase 2 must be aware of all mapping updates.
```

Comment:
More context on this: 
After first phase (collect schema updates), Indexer's target columns can be updated in 2 ways
1. New columns are added to the schema == some target columns got OID-s. We will receive and use them in a flow up.
2. Existing object columns can get some new sub-columns. Reusing instance is just an easier way to make phase 1's parent object column to be aware of the new sub-columns. Otherwise we need to find a way to replace ALL target refs after performing `addColumnAction` which might be tricky.